src/typechecker.rs

```diff
@@ -298,6 +349,79 @@ impl<'a> Typechecker<'a> {
                 block,
             } => self.typecheck_def(name, params, return_ty, block, node_id),
             AstNode::Call { ref parts } => self.typecheck_call(parts, node_id),
+            AstNode::For {
+                variable,
+                range,
+                block,
+            } => {
+                self.typecheck_node(range);
+
+                // We don't need to typecheck variable after this
+                let var_id = self
+                    .compiler
+                    .var_resolution
+                    .get(&variable)
+                    .expect("missing resolved variable");
+                if let Type::List(type_id) = self.type_of(range) {
+                    self.variable_types[var_id.0] = type_id;
+                    self.set_node_type_id(variable, type_id);
+                } else {
+                    self.variable_types[var_id.0] = ANY_TYPE;
+                    self.set_node_type_id(variable, ERROR_TYPE);
+                    self.error("For loop range is not a list", range);
+                }
+
+                self.typecheck_node(block);
+                if self.type_id_of(block) != NONE_TYPE {
+                    self.error("Blocks in looping constructs cannot return values", block);
+                }
+
+                self.set_node_type_id(node_id, NOTHING_TYPE);
```

Comment:
To match the current behavior of Nushell (i.e., `for i in 0..8 { print $i } | describe` throwing an error because `for` is a statement), we should probably make it into a `none` type.

However, in the future, we could perhaps allow `for` to be an expression, and you'd be able to break a value from it, e.g., `let x = for i in 1..8 { if $ > 3 { break ($i * $i) } }`. In that case, `nothing` type is appropriate for a loop without value breaks. This applies to `while` and `loop` as well.

Reply:
For the second part, would it not be a problem if the for loop never evaluates with a list at runtime? Or do you plan to keep it `oneof<nothing, type>`?

Reply:
For now I've gone with `none`